src/core/json/driver.cc

```diff
@@ -27,4 +89,22 @@ void Driver::ResetScanner() {
   lexer_ = make_unique<Lexer>();
 }
 
+void Driver::AddFunction(string_view fname) {
+  if (!path_.empty()) {
+    throw Parser::syntax_error(lexer_->location(),
+                               "function can be only at the beginning of the path");
+  }
+  DCHECK(!agg_);
+  if (fname == "max") {
+    agg_.reset(new MaxImpl);
+  } else if (fname == "min") {
+    agg_.reset(new MinImpl);
+  } else if (fname == "avg") {
+    agg_.reset(new AvgImpl);
+  } else {
```

Comment:
suggestion
```
static map<string, std::function<AggFunction*()>> functors{
      {"min", [] { return new MinImpl(); }}, 
      {"max", [] { return new MaxImpl(); }}
};
return unique_ptr(functors[name]()) // and check error
```

Reply:
my subjective preference is to old conditionals . but if you have a suggestion on how to simplify the the traversal/functor code, would love to hear

Reply:
you can't have a function pointer to a constructor